File: July2022/Python/constructBinaryTreeFromPreorderAndInorderTraversal.py
@Jiganesh

# https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution(object):
    
    # Runtime: 115 ms, faster than 75.80% of Python online submissions for Construct Binary Tree from Preorder and Inorder Traversal.
    # Memory Usage: 19 MB, less than 73.98% of Python online submissions for Construct Binary Tree from Preorder and Inorder Traversal.
    
    # Efficient and Optimized
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        index_lookup =  {val:ind for ind, val in enumerate (inorder)}
        def helper(instart, inend , prestart, preend):
            
            if prestart > preend or instart>inend : return None
            
            root = TreeNode(preorder[prestart])
            
            root_in_index = index_lookup[preorder[prestart]]

            element_to_left = root_in_index-instart
            
            root.left = helper(instart, root_in_index-1 ,prestart+1, prestart+element_to_left+1)
            root.right = helper(root_in_index+1, inend , prestart+element_to_left+1, preend)
            
            return root
    
        return helper(0, len(inorder)-1, 0, len(preorder)-1,)
    
    # Extra Space but Elegant Code
    
    # Runtime: 129 ms, faster than 72.91% of Python online submissions for Construct Binary Tree from Preorder and Inorder Traversal.
    # Memory Usage: 52.3 MB, less than 55.77% of Python online submissions for Construct Binary Tree from Preorder and Inorder Traversal.    
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        

        def helper (inorder):
            root =None
            if inorder :
                
                root_val = preorder.pop(0)
                root_ind = inorder.index(root_val)
                
                root = TreeNode(root_val)
                
                root.left = helper(inorder[:root_ind])
                root.right = helper(inorder[root_ind+1:])
                
            return root
        
        return helper(inorder)
    
    # More Elegant 
    def buildTree(self, preorder, inorder):
        if inorder:
            INDEX = inorder.index(preorder.pop(0))
            root = TreeNode(inorder[INDEX])
            root.left = self.buildTree(preorder, inorder[:INDEX])
            root.right = self.buildTree(preorder, inorder[INDEX+1:])
			
            return root

File: July2022/Java/ConstructBinaryTreeFromPreorderAndInorderTraversal.java

// @saorav21994
// TC : O(n)
// SC : O(n)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    HashMap<Integer, Integer> indexMap = new HashMap<>();
    int preCtr = 0;
    public TreeNode computeTree(int [] preorder, int [] inorder, int start, int end) {
        if (start > end) 
            return null;   // no node exits
        TreeNode node = new TreeNode(preorder[preCtr]);
        preCtr += 1;
        if (start == end)
            return node;  // only sole node
        
        // search index of the node in hashMap
        int i = indexMap.get(node.val);
        node.left = computeTree(preorder, inorder, start, i-1);
        node.right = computeTree(preorder, inorder, i+1, end);
        
        return node;
    }
    
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        for (int i = 0; i < inorder.length; i++) {
            indexMap.put(inorder[i], i);
        }
        TreeNode root = computeTree(preorder, inorder, 0, preorder.length-1);
        return root;
    }
}
